### DISCUSSION TITLE
Band diagram of 2D structure periodic in only one direction

### DISCUSSION THREAD

**Discussion by peterropac** [NanoComp/meep#2840] [Q&A] [answered]
I would like to calculate the band diagram of a 2D structure, but only for k-vectors pointing in one of the principal directions. The structure is also only periodic in that same direction. What would be the best way to do that? Is it possible to do it with MPB?

**Comment by stevengj** [comment]
If it's only periodic in one direction, what does it look like in the other direction?  A picture of your structure would help.

**Reply by peterropac** [reply]
It is a waveguide extending in the x direction but it's width (y axis) is periodically changing. It is symmetric with respect to the x-axis. If I simulated it in Meep i would put PMLs in the y-axis. My "solution" was to extend the unit cell along the y-axis so the majority is empty space. I wonder if there is a better solution.

**Comment by dalarev** [comment]
Define your [k_points](https://mpb.readthedocs.io/en/stable/Python_User_Interface/#the-modesolver-class) only along the direction of interest.

**Reply by peterropac** [reply]
I did do that, but as far as I understand the cell is still technically periodic along both axes.

**Comment by stevengj** [comment, marked as answer]
> It is a waveguide extending in the x direction but it's width (y axis) is periodically changing.

If you are just interested in the guided modes (i.e. under the light line, see [chapter 6 of our book](http://ab-initio.mit.edu/book/)), then you can just use a supercell in the y direction.  Since the guided modes are exponentially localized, then the boundary conditions in y won't affect them if the cell size is sufficiently large.

See, for example, the "2d periodic waveguide" example in [this tutorial](https://nbviewer.org/github/mitmath/18369/blob/master/notes/MPB-demo.ipynb)

**Reply by peterropac** [reply]
Ok, thank you! That is along the lines of what I was thinking.
